Clock: adding support for `repr` and `str` functionality (#2829)
exercism/python: exercises/practice/clock/.docs/instructions.append.md

# Instructions append

## Representing your class

When working with and debugging objects, it's important to have a string representation of that object. For example, if you were to create a `datetime.datetime` object in a REPL environment, you could see its string representation:

```python
>>> from datetime import datetime
>>> datetime(2022, 5, 4)
datetime.datetime(2022, 5, 4, 0, 0)
```

In this excersize, you will create a custom object, `Clock`, that handles times without dates. To do this in python, you would create a `class` to define its data and behavior.

One aspect of a class' behavior is this string prepresentation. By default, a class' string representation is not very helpful:

```python
>>> Clock(12, 34)

> <Clock object st 0x... >

Author: Sounds great! Added.

```

To create a more helpful representation, you will need to define a `__repr__` method on your class that returns a string.

> According to the [specification for a `__repr__` method](https://docs.python.org/3/reference/datamodel.html#object.__repr__), you should ideally have the method return a string that could be used to recreate the object, if you were to copy and paste it into a new environment. 

Reviewer: 1. Prefer reference links, per https://exercism.org/docs/building/markdown/markdown
2.

```suggestion
According to the [specification for a `__repr__` method](https://docs.python.org/3/reference/datamodel.html#object.__repr__), you should ideally have the method return valid Python code that could be used to recreate the object.
That allows you to copy-paste the output directly into code or the REPL.
```

> For example, a `Clock` that represents 11:30 AM could look like this: `Clock(11, 30)`.
> 
> 

Reviewer: ```suggestion
For example, a `Clock` that represents 11:30 AM could look like this: `Clock(11, 30)`.

```

> Defining a `__repr__` method is a great thing to do for every class you create. When you do, you should consider:

Reviewer: ```suggestion
Defining a `__repr__` method is good practice for all classes.
When you do, you should consider:
```

> 
> - The information you're returning from this method will be beneficial when you're debugging issues.
> 
> - Ideally, you should return a string that a user could copy, paste into a python environment, and have a copy of the object.
> 
> - If that's not practical, return a description between angle brackets, as in `< ...a practical description... >`.

Reviewer: ```suggestion

- The information you're returning from this method will be beneficial when you're debugging issues.
- Ideally, you should return a string that is valid Python code.
- If that is not practical, return a description between angle brackets: `< ...a practical description... >`.
```

### String conversion

> As useful as the `__repr__` method is, sometimes you'll want an alternative string representation. For example, let's look at the `datetime.datetime` class again:

Reviewer: ```suggestion
In addition to the `__repr__` method, sometimes you'll want an alternative string representation.
This might be used to format the object for program output.
For example, let's look at the `datetime.datetime` class again:
```

```python
>>> str(datetime.datetime(2022, 5, 4))
'2022-05-04 00:00:00'
```

> When a `datetime` object is askes to convert itself to a string, it turns into a string formatted according to the ISO 8601 standard, which can be parsed by most datetime libraries.

Reviewer: ```suggestion
When a `datetime` object is askes to convert itself to a string, it returns a string formatted according to the ISO 8601 standard, which can be parsed by most datetime libraries.
```

> This clock exercise will provide you with a similar opportunity. Part of the challenge will be to have an alternate representation like this:

Reviewer: ```suggestion
In this exercise, you will get to write a `__str__` method, too.
```

# Instructions append

> 

Reviewer 2: ```suggestion

The tests for this exercise expect your clock will be implemented via a Clock `class` in Python.
If you are unfamiliar with classes in Python, [classes][classes in python] from the Python docs is a good place to start.
```

## Representing your class

When working with and debugging objects, it's important to have a string representation of that object.

> For example, if you were to create a `datetime.datetime` object in a REPL environment, you can see its string representation:

Reviewer 2: ```suggestion
For example, if you were to create a new `datetime.datetime` object in the Python [REPL][REPL] environment, you could see its string representation:
```

> ```python
> >>> from datetime import datetime
> >>> datetime(2022, 5, 4)
> datetime.datetime(2022, 5, 4, 0, 0)
> ```

Reviewer 2: ```suggestion
```python
>>> from datetime import datetime
>>> new_date = datetime(2022, 5, 4)
>>> new_date
datetime.datetime(2022, 5, 4, 0, 0)
```
```

> In this excersise, you will create a custom object, `Clock`, that handles times without dates.
> To do this in Python, you would create a `class` to to implement the desired behavior.
> 
> One aspect of a class's behavior is how it is represented as a string.
> By default, a class's string representation is not very helpful.

Reviewer 2: ```suggestion
Your Clock `class` will create a custom object that handles times without dates.

One important aspect of this `class` will be how it is represented as a string.
Other programmers who use or call Clock objects created from the Clock `class` will refer to this string representation for debugging and  other activities.  
However, by default a class's string representation is not very helpful.
```

```python
>>> Clock(12, 34)
<Clock object st 0x102807b20 >
```

> To create a more helpful representation, you need to define a `__repr__` method on your class that returns a string.

Reviewer 2: ```suggestion
To create a more helpful representation, you can define a `__repr__` method on the class that returns a string.
```

> According to the [specification for a `__repr__` method][repr-docs], you should ideally have the method return valid Python code that could be used to recreate the object.
> That allows you to copy-paste the output directly into code or the REPL.
> 

Reviewer 2: ```suggestion
Ideally, the `__repr__` method returns valid Python code that can be used to recreate the object -- as laid out in the [specification for a `__repr__` method][repr-docs].
Returning valid Python code allows you to copy-paste the `str` directly into code or the REPL.

```

For example, a `Clock` that represents 11:30 AM could look like this: `Clock(11, 30)`.

Defining a `__repr__` method is good practice for all classes.

> When you do, you should consider:

Reviewer 2: ```suggestion
Some things to consider:
```

> - The information you're returning from this method will be beneficial when you're debugging issues.

Reviewer 2: ```suggestion
- The information returned from this method should be beneficial when debugging issues.
```

> - Ideally, you should return a string that is valid Python code.

Reviewer 2: ```suggestion
- _Ideally_, the method returns a string that is valid Python code -- although that might not always be possible.
```

> - If that is not practical, return a description between angle brackets: `< ...a practical description... >`.

Reviewer 2: ```suggestion
- If valid Python code is not practical, returning a description between angle brackets: `< ...a practical description... >` is the convention.
```

### String conversion

> In addition to the `__repr__` method, sometimes you'll want an alternative string representation.

Reviewer 2: ```suggestion
In addition to the `__repr__` method, there might also be a need for an alternative "human-readable" string representation of the `class`.
```

> This might be used to format the object for program output.

Reviewer 2: ```suggestion
This might be used to format the object for program output, or documentation.
```

> For example, let's look at the `datetime.datetime` class again:

Reviewer 2: ```suggestion
Looking at `datetime.datetime` again:
```

```python
>>> str(datetime.datetime(2022, 5, 4))
'2022-05-04 00:00:00'
```

> When a `datetime` object is askes to convert itself to a string, it returns a string formatted according to the ISO 8601 standard, which can be parsed by most datetime libraries.

Reviewer 2: ```suggestion
When a `datetime` object is asked to convert itself to a string representation, it returns a `str` formatted according to the [ISO 8601 standard][ISO 8601], which can be parsed by most datetime libraries into a human-readable date and time.
```

> In this exercise, you will get to write a `__str__` method, too.

Reviewer 2: ```suggestion
In this exercise, you will get a chance to write a `__str__` method, as well as a `__repr__`.
```

```python
>>> str(Clock(11, 30))
'11:30'
```

To support this string conversion, you will need to create a `__str__` method on your class.

> [repr-docs]: https://docs.python.org/3/reference/datamodel.html#object.__repr__

Reviewer 2: ```suggestion
[repr-docs]: https://docs.python.org/3/reference/datamodel.html#object.__repr__
[classes in python]: https://docs.python.org/3/tutorial/classes.html
[REPL]: https://pythonprogramminglanguage.com/repl/
[ISO 8601]: https://www.iso.org/iso-8601-date-and-time-format.html

```

For example, a `Clock` that represents 11:30 AM could look like this: `Clock(11, 30)`.

Defining a `__repr__` method is good practice for all classes.
Some things to consider:

- The information returned from this method should be beneficial when debugging issues.
- _Ideally_, the method returns a string that is valid Python code -- although that might not always be possible.
- If valid Python code is not practical, returning a description between angle brackets: `< ...a practical description... >` is the convention.

### String conversion

In addition to the `__repr__` method, there might also be a need for an alternative "human-readable" string representation of the `class`.
This might be used to format the object for program output or documentation.
Looking at `datetime.datetime` again:

```python
>>> str(datetime.datetime(2022, 5, 4))
'2022-05-04 00:00:00'
```

When a `datetime` object is asked to convert itself to a string representation, it returns a `str` formatted according to the [ISO 8601 standard][ISO 8601], which can be parsed by most datetime libraries into a human-readable date and time.

> In this exercise, you will get a chance to write a `__str__` method, as well as a `__repr__`.

Reviewer 2: ```suggestion
In this exercise, you will get a chance to write a `__str__` method for "human-readable" representation, as well as a `__repr__` method that returns Python code.
```

```python
>>> str(Clock(11, 30))
'11:30'
```

> To support this string conversion, you will need to create a `__str__` method on your class.

Reviewer 2: ```suggestion
To support this string conversion, you will need to create a `__str__` method on your class that returns a more "human readable" string showing the Clock time.
```

[repr-docs]: https://docs.python.org/3/reference/datamodel.html#object.__repr__
[classes in python]: https://docs.python.org/3/tutorial/classes.html
[REPL]: https://pythonprogramminglanguage.com/repl/
[ISO 8601]: https://www.iso.org/iso-8601-date-and-time-format.html

> 

Reviewer 2: Not sure if I am overloading this too much -- thinking we _might_ want a document/link here going over `__str__` vs `__repr__` in more detail?  Then again, most I've seen just confuse the issue.  What are your thoughts?

Author: That could be a great idea. I have a hard time finding something that felt right to me, but [this stack overflow answer](https://stackoverflow.com/a/28132458/496097) seems like one of the best descriptions I've seen.

And it reminds me that I want to add something about how python uses `__repr__` to represent an object if `__str__` isn't present.

Reviewer 2: >And it reminds me that I want to add something about how python uses __repr__ to represent an object if __str__ isn't present.

_ooh._ YES.  Do you want to take a shot at adding that in here, or do a follow-on PR?  Either works for me.  😄 

And I agree - I have yet to find something self-contained, readable, and concise.  (_so we probably should make one....!!_) But that SO link is about as close as I've seen.
